### ISSUE TITLE
Support model chaining across Predictor types

### ISSUE THREAD

**Issue by deliahu** [cortexlabs/cortex#821]
#### Description

Provide a simple way to chain models together across predictor types, e.g. GPT2 (TensorFlow Predictor) -> Tacotron (Python Predictor).

This would be a single API that runs in the cluster, which makes requests to both other APIs (which may not be exposed to the internet)

**Comment by tthebst** [comment]
I gave this some thought because it sounds useful. 
One possible option would be to create another predictor for example Chaining preditor:

`
labels = ["setosa", "versicolor", "virginica"]

class ChainPredictor:
    def __init__(self, chain_client, config):
        self.client = chain_client
        self.models =["GPT","Tacotron"]

    def predict(self, payload):
        model_input = [
            payload["sepal_length"],
            payload["sepal_width"],
            payload["petal_length"],
            payload["petal_width"],
        ]

        prediction = self.chain.predict(model_input)
        predicted_class_id = prediction[0][0]
        return labels[predicted_class_id]
`

The `self.models` would then contain the names of already deployed models and one would ten implement the calls to all models in the `chain_client.predict `.

One possible addition would also be to add transformation function to apply between models.

Tim

**Comment by deliahu** [comment]
@tthebst thanks for reaching out on this! @vishalbollu and I just discussed, and we think that for now, it's easy enough to use multiple models in a single endpoint for the Python predictor; you can load both models in `__init__()`, and then call them sequentially in `predict()`.

We do not yet support this for the TensorFlow Predictor type (you'd have to use the Python Predictor with TensorFlow models). I just updated the description on https://github.com/cortexlabs/cortex/issues/890 to be more clear, which now covers this topic.

Therefore, since we think it's supported well enough for the Python Predictor, and we now have an issue to track TensorFlow/ONNX, I'm thinking of closing this issue, what do you think? Let me know if we're missing a big opportunity here to improve it though!

**Comment by tthebst** [comment]
Yes, that's true, but what if someone wants to mix TensorFlow with python predictor. For example, if one could use pretrained tf models, which outputs then get used in a for example python model. I don't know how common this workflow is but as of now one would have to stick to one of the deployment options to do model chaining. Or implement the chaining on the client-side.

**Comment by deliahu** [comment]
@tthebst Yes you are correct; to chain models across providers, users will have to implement it on the client.

Since we have not yet heard a feature request for this, and since there is a reasonable work-around which makes it possible (implementing it on the client), I think for now supporting server-side model chaining across predictor types is not a high priority, do you think that makes sense?

**Comment by deliahu** [comment]
@tthebst we've decided to not close this ticket, but rather update the title/description and de-prioritize it for now

**Comment by deliahu** [comment]
Since we now support multi-model TensorFlow APIs, it is possible to chain multiple TensorFlow models in a single predictor.

In addition, it is possible to chain models of multiple different types (TensorFlow, Python, or ONNX) this by using a "wrapper" API. This wrapper API would be a [Python Predictor type](https://docs.cortex.dev/deployments/predictors#python-predictor), and would make prediction requests and perform the reduction in the `predict()` function. All inference APIs would autoscale independently.

If anyone has questions about either of these approaches, we can provide an example implementation.
